### PR TITLE
Add new article: Inspecting assembly contents using MetadataLoadContext

### LINKED ISSUE #17360: New article and sample idea: Inspecting assembly contents using MetadataLoadContext
.NET Framework had the reflection-only assembly load mode that allowed to inspect the contents of an assembly without loading it for execution. This mode is not supported in .NET Core. The replacement API is [MetadataLoadContext](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.metadataloadcontext?view=dotnet-plat-ext-3.1), however, there's no information how to start using it. Searching for "c# MetadataLoadContext" only gives API reference and some unanswered StackOverflow questions. I suggest to create article similar to https://docs.microsoft.com/en-us/dotnet/framework/reflection-and-codedom/how-to-load-assemblies-into-the-reflection-only-context that will cover MetadataLoadContext

## Suggested TOC location
Under https://docs.microsoft.com/en-us/dotnet/standard/assembly/

## Abstract
 The article will explain how to create the isolated MetadataLoadContext using fixed set of paths to resolve assemblies, load assembly into it and extract information about metdata (for example, attributes, defined types...)

## Outline
- Introduction: purpose of MetadataLoadContext, possible scenarios when it is useful
- Prerequisites
- Create MetadataAssemblyResolver for MetadataLoadContext
- Create MetadataLoadContext
- Example

My suggested code sample that can be used for this article: https://github.com/MSDN-WhiteKnight/CodeSamples/tree/master/MetadataLoadContextSample

### PR DESCRIPTION
**Pull request by MSDN-WhiteKnight**
Fixes #17360

Contributes to #2729

Related samples PR: https://github.com/dotnet/samples/pull/2053

### CODE REVIEW

**Comment by Thraka** [comment]
@MSDN-WhiteKnight Looking good! You should move your snippet content into a child folder named for the article. This way others can include snippets for other articles in the assembly area.

**Review by Thraka — CHANGES_REQUESTED** [review]
Per my other comment. The snippet should be in its own folder

File: docs/standard/assembly/inspect-contents-using-metadataloadcontext.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+## Create MetadataAssemblyResolver for MetadataLoadContext
+
+Creating the <xref:System.Reflection.MetadataLoadContext> requires providing the instance of the <xref:System.Reflection.MetadataAssemblyResolver>. The simplest way to provide one is to use the <xref:System.Reflection.PathAssemblyResolver>, which resolves assemblies from the given collection of assembly path strings. This collection, besides assemblies you want to inspect directly, should also include all needed dependencies. For example, to read the custom attribute located in an external assembly, you should include that assembly, or an exception will be thrown. In most cases, you should include at least the *core assembly*, that is, the assembly containing built-in system types, such as <xref:System.Object?displayProperty=nameWithType>. The following code shows how to create the <xref:System.Reflection.PathAssemblyResolver> using the collection consisting of the inspected assembly and the current runtime's core assembly:
```
Comment (gewarren) [inline docs/standard/assembly/inspect-contents-using-metadataloadcontext.md]:
```suggestion
Creating the <xref:System.Reflection.MetadataLoadContext> requires providing the instance of the <xref:System.Reflection.MetadataAssemblyResolver>. The simplest way to provide one is to use the <xref:System.Reflection.PathAssemblyResolver>, which resolves assemblies from the given collection of assembly path strings. This collection, besides assemblies you want to inspect directly, should also include all needed dependencies. For example, to read the custom attribute located in an external assembly, you should include that assembly or an exception will be thrown. In most cases, you should include at least the *core assembly*, that is, the assembly containing built-in system types, such as <xref:System.Object?displayProperty=nameWithType>. The following code shows how to create the <xref:System.Reflection.PathAssemblyResolver> using the collection consisting of the inspected assembly and the current runtime's core assembly:
```

File: docs/standard/assembly/inspect-contents-using-metadataloadcontext.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+[!code-csharp[](snippets/inspect-contents-using-metadataloadcontext/MetadataLoadContextSnippets.cs#CoreAssembly)]
+
+If you need access to all BCL types, you can include all runtime assemblies into the collection. The following code shows how to create the <xref:System.Reflection.PathAssemblyResolver> using the collection consisting of the inspected assembly and all assemblies of the current runtime:
```
Comment (gewarren) [inline docs/standard/assembly/inspect-contents-using-metadataloadcontext.md]:
```suggestion
If you need access to all BCL types, you can include all runtime assemblies in the collection. The following code shows how to create the <xref:System.Reflection.PathAssemblyResolver> using the collection consisting of the inspected assembly and all assemblies of the current runtime:
```

File: docs/standard/assembly/inspect-contents-using-metadataloadcontext.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+## Create MetadataLoadContext
+
+To create the <xref:System.Reflection.MetadataLoadContext>, invoke its constructor <xref:System.Reflection.MetadataLoadContext.%23ctor%28System.Reflection.MetadataAssemblyResolver%2CSystem.String%29>, passing the previously created <xref:System.Reflection.MetadataAssemblyResolver> as the first parameter and the core assembly name as the second parameter. You can omit the core assembly name, in which case the constructor will attempt to use default names: "mscorlib", "System.Runtime", or "netstandard".
+
+After you've created the context, you can load assemblies into it using methods such as <xref:System.Reflection.MetadataLoadContext.LoadFromAssemblyPath%2A>. You can use all reflection APIs on loaded assemblies, except ones that involve code execution. Note that the <xref:System.Reflection.MemberInfo.GetCustomAttributes%2A> method does involve the execution of constructors, so you should use the <xref:System.Reflection.MemberInfo.GetCustomAttributesData%2A> method instead when you need to examine custom attributes in the <xref:System.Reflection.MetadataLoadContext>.
```
Comment (gewarren) [inline docs/standard/assembly/inspect-contents-using-metadataloadcontext.md]:
```suggestion
After you've created the context, you can load assemblies into it using methods such as <xref:System.Reflection.MetadataLoadContext.LoadFromAssemblyPath%2A>. You can use all reflection APIs on loaded assemblies except ones that involve code execution. The <xref:System.Reflection.MemberInfo.GetCustomAttributes%2A> method does involve the execution of constructors, so use the <xref:System.Reflection.MemberInfo.GetCustomAttributesData%2A> method instead when you need to examine custom attributes in the <xref:System.Reflection.MetadataLoadContext>.
```

File: docs/standard/assembly/inspect-contents-using-metadataloadcontext.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+The following code sample creates <xref:System.Reflection.MetadataLoadContext>, loads the assembly into it and outputs assembly attributes into the console:
```
Comment (gewarren) [inline docs/standard/assembly/inspect-contents-using-metadataloadcontext.md]:
```suggestion
The following code sample creates <xref:System.Reflection.MetadataLoadContext>, loads the assembly into it, and outputs assembly attributes into the console:
```

**Review by gewarren — COMMENTED** [review]
Looks good with some minor suggestions. Thanks!

**Comment by gewarren** [comment]
@Thraka The snippet is in its own folder now AFAICT. Can you approve if you agree?

**Review by Thraka — APPROVED** [review]

**Pull request merged**